### PR TITLE
Feat/adding device id to fxa data

### PR DESCRIPTION
**Pull request by kik-kik**
Deriving from PR https://github.com/mozilla/bigquery-etl/pull/3234

The reason is to break down the changes into smaller chunks and deploy the changes step-by-step to reduce changes of errors.

### CODE REVIEW

**Comment by kik-kik** [comment]
@relud quick questions around bqetl:
- Looking at the `metadata.yaml` files for these models they appear to have: `arguments: ['--schema_update_option=ALLOW_FIELD_ADDITION']`

Could you confirm in this case I do not need to make a `v2` of these models. Adjusting the query to include a new field should work just fine here and automatically should be added to the schema? After that I can run `backfill` cmd on this model for the last 30 days (only how long the original data is retained) without any impact on this table and the data in it? This would mean that simple query adjustment should be enough here without multiple versions and having to union them together in the view?

Second question (just for the future), if I delete one of the models in bqetl nothing happens, right? It will remain in BQ unless manually deleted? Just wondering if I ever need to introduce a new version of a model do I have to retain both folders or if just the most recent one.

Thanks!

**Comment by kik-kik** [comment]
Frank already provided me with answers.

**Review by fbertsch — COMMENTED** [review]
@kik-kik what is the `device_id`?

1. We hash `user_id` to keep it private, do we need to do the same for `device_id`?
2. Can you add a comment describing what the field is? Not sure if there's docs to link to

**Comment by kik-kik** [comment]
Based on my understanding from previous conversations `device_id` is a hashed value already. Something mentioned on slack: `I believe that if you disconnect a device from an account and reconnect it, you won't get the same device ID` -> to me this sounds fine.

**Comment by kik-kik** [comment]
2. Here's some docs I got:
https://github.com/mozilla/fxa/blob/main/packages/fxa-auth-server/docs/metrics-events.md

One of the definition: ```Optional. The id of the device record. This does correlate back to a record the FxA user db.```

The other literally just says it's active device id.

File: sql/moz-fx-data-shared-prod/firefox_accounts_derived/fxa_auth_events_v1/query.sql [hunk 1]
```diff
@@ -10,7 +10,7 @@ WITH base AS (
           jsonPayload.* REPLACE (
             (
               SELECT AS STRUCT
-                jsonPayload.fields.* EXCEPT (device_id, user_id) REPLACE(
+                jsonPayload.fields.* EXCEPT (user_id) REPLACE(
                   -- See https://bugzilla.mozilla.org/show_bug.cgi?id=1707571
```
Comment (relud) [inline sql/moz-fx-data-shared-prod/firefox_accounts_derived/fxa_auth_events_v1/query.sql:14]: add `TO_HEX(SHA256(jsonPayload.fields.device_id)) AS device_id,` after `user_id` 8 lines below and:
```suggestion
                jsonPayload.fields.* EXCEPT (device_id, user_id) REPLACE(
                  -- See https://bugzilla.mozilla.org/show_bug.cgi?id=1707571
```

File: sql/moz-fx-data-shared-prod/firefox_accounts_derived/fxa_content_events_v1/query.sql [hunk 2]
```diff
@@ -5,7 +5,7 @@ SELECT
         jsonPayload.* REPLACE (
           (
             SELECT AS STRUCT
-              jsonPayload.fields.* EXCEPT (device_id, user_id),
+              jsonPayload.fields.* EXCEPT (user_id),
               TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id
```
Comment (relud) [inline sql/moz-fx-data-shared-prod/firefox_accounts_derived/fxa_content_events_v1/query.sql:9]:
```suggestion
              jsonPayload.fields.* EXCEPT (device_id, user_id),
              TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id,
              TO_HEX(SHA256(jsonPayload.fields.device_id)) AS device_id,
```

File: sql/moz-fx-data-shared-prod/firefox_accounts_derived/fxa_stdout_events_v1/query.sql [hunk 3]
```diff
@@ -5,7 +5,7 @@ SELECT
         jsonPayload.* REPLACE (
           (
             SELECT AS STRUCT
-              jsonPayload.fields.* EXCEPT (device_id, user_id),
+              jsonPayload.fields.* EXCEPT (user_id),
               TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id
```
Comment (relud) [inline sql/moz-fx-data-shared-prod/firefox_accounts_derived/fxa_stdout_events_v1/query.sql:9]:
```suggestion
              jsonPayload.fields.* EXCEPT (device_id, user_id),
              TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id,
              TO_HEX(SHA256(jsonPayload.fields.device_id)) AS device_id,
```

File: sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_content_events_v1/query.sql [hunk 4]
```diff
@@ -5,7 +5,7 @@ SELECT
         jsonPayload.* REPLACE (
           (
             SELECT AS STRUCT
-              jsonPayload.fields.* EXCEPT (device_id, user_id),
+              jsonPayload.fields.* EXCEPT (user_id),
               TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id
```
Comment (relud) [inline sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_content_events_v1/query.sql:9]:
```suggestion
              jsonPayload.fields.* EXCEPT (device_id, user_id),
              TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id,
              TO_HEX(SHA256(jsonPayload.fields.device_id)) AS device_id,
```

File: sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_stdout_events_v1/query.sql [hunk 5]
```diff
@@ -5,7 +5,7 @@ SELECT
         jsonPayload.* REPLACE (
           (
             SELECT AS STRUCT
-              jsonPayload.fields.* EXCEPT (device_id, user_id),
+              jsonPayload.fields.* EXCEPT (user_id),
               TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id
```
Comment (relud) [inline sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_stdout_events_v1/query.sql:9]:
```suggestion
              jsonPayload.fields.* EXCEPT (device_id, user_id),
              TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id,
              TO_HEX(SHA256(jsonPayload.fields.device_id)) AS device_id,
```

File: sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_auth_events_v1/query.sql [hunk 6]
```diff
@@ -5,15 +5,16 @@ SELECT
...
                 -- See https://bugzilla.mozilla.org/show_bug.cgi?id=1707571
                 CAST(NULL AS FLOAT64) AS emailverified,
                 CAST(NULL AS FLOAT64) AS isprimary,
                 CAST(NULL AS FLOAT64) AS isverified,
                 -- casting id as field type in source tables inconsistent
                 CAST(jsonPayload.fields.id AS STRING) AS id
               ),
-              TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id
+              TO_HEX(SHA256(jsonPayload.fields.user_id)) AS user_id,
+              COALESCE(jsonPayload.fields.device_id, jsonPayload.fields.deviceid) AS device_id
```
Comment (relud) [inline sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_auth_events_v1/query.sql:17]:
```suggestion
              TO_HEX(SHA256(COALESCE(jsonPayload.fields.device_id, jsonPayload.fields.deviceid))) AS device_id,
```

Comment (relud) [inline sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_auth_events_v1/query.sql:17]: why are we only fixing `device_id` and `deviceid` in nonprod?

**Review by relud — APPROVED** [review]
r+wc

> We hash `user_id` to keep it private, do we need to do the same for `device_id`?

I believe we do. `user_id` and `device_id` are both 32-character hex strings before we hash them, and would exactly match an index in the FxA db, and I believe that is what we hash `user_id` to prevent, so we should do it with `device_id` as well.

**Review by gkabbz — APPROVED** [review]

File: sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_auth_events_v1/query.sql (hunk 6 above)
Comment (kik-kik) [reply sql/moz-fx-data-shared-prod/firefox_accounts_derived/nonprod_fxa_auth_events_v1/query.sql:17]: good point, I will make that tweak now.

**Review by relud — APPROVED** [review]

**Comment by kik-kik** [comment]
@fbertsch I created a PR to add a little bit more context around the field to the `fxa` docs:
https://github.com/mozilla/fxa/pull/14459/files

Please let me know if this is sufficient. Thanks!

**Review by fbertsch — APPROVED** [review]
> @fbertsch I created a PR to add a little bit more context around the field to the `fxa` docs: https://github.com/mozilla/fxa/pull/14459/files
> 
> Please let me know if this is sufficient. Thanks!

LGTM!

**Comment by whd** [comment]
I'm seeing the following on views deploys since this was merged:

```yaml
configuration:
  dryRun: false
  jobType: QUERY
  query:
    destinationTable:
      datasetId: firefox_accounts
      projectId: moz-fx-data-shared-prod
      tableId: fxa_content_auth_oauth_events
    priority: INTERACTIVE
    query: |
      CREATE OR REPLACE VIEW
        `moz-fx-data-shared-prod.firefox_accounts.fxa_content_auth_oauth_events`
      AS
        --
      WITH content AS (
        SELECT
          jsonPayload.logger,
          jsonPayload.fields.event_type,
          jsonPayload.fields.app_version,
          jsonPayload.fields.os_name,
          jsonPayload.fields.os_version,
          jsonPayload.fields.country,
          jsonPayload.fields.language,
          jsonPayload.fields.user_id,
          jsonPayload.fields.user_properties,
          jsonPayload.fields.event_properties,
          jsonPayload.fields.device_id,
          `timestamp`,
          receiveTimestamp,
        FROM
          `moz-fx-data-shared-prod.firefox_accounts_derived.fxa_content_events_v1`
      ),
          --
      auth AS (
        SELECT
          jsonPayload.logger,
          jsonPayload.fields.event_type,
          jsonPayload.fields.app_version,
          jsonPayload.fields.os_name,
          jsonPayload.fields.os_version,
          jsonPayload.fields.country,
          jsonPayload.fields.language,
          jsonPayload.fields.user_id,
          jsonPayload.fields.user_properties,
          jsonPayload.fields.event_properties,
          jsonPayload.fields.device_id,
          `timestamp`,
          receiveTimestamp,
        FROM
          `moz-fx-data-shared-prod.firefox_accounts_derived.fxa_auth_events_v1`
      ),
          --
      oauth AS (
        SELECT
          jsonPayload.logger,
          jsonPayload.fields.event_type,
          jsonPayload.fields.app_version,
          CAST(NULL AS STRING) AS os_name,
          CAST(NULL AS STRING) AS os_version,
          CAST(NULL AS STRING) AS country,
          CAST(NULL AS STRING) AS language,
          jsonPayload.fields.user_id,
          jsonPayload.fields.user_properties,
          jsonPayload.fields.event_properties,
          CAST(NULL AS STRING) AS device_id,
          `timestamp`,
          receiveTimestamp,
        FROM
          `moz-fx-data-shared-prod.firefox_accounts_derived.fxa_oauth_events_v1`
      ),
          --
      unioned AS (
        SELECT
          *
        FROM
          auth
        UNION ALL
        SELECT
          *
        FROM
          content
        UNION ALL
        SELECT
          *
        FROM
          oauth
      )
          --
      SELECT
        * EXCEPT (user_properties, event_properties),
        JSON_VALUE(user_properties, "$.utm_term") AS utm_term,
        JSON_VALUE(user_properties, "$.utm_source") AS utm_source,
        JSON_VALUE(user_properties, "$.utm_medium") AS utm_medium,
        JSON_VALUE(user_properties, "$.utm_campaign") AS utm_campaign,
        JSON_VALUE(user_properties, "$.utm_content") AS utm_content,
        JSON_VALUE(user_properties, "$.ua_version") AS ua_version,
        JSON_VALUE(user_properties, "$.ua_browser") AS ua_browser,
        JSON_VALUE(user_properties, "$.entrypoint") AS entrypoint,
        JSON_VALUE(user_properties, "$.flow_id") AS flow_id,
        JSON_VALUE(event_properties, "$.service") AS service,
        JSON_VALUE(event_properties, "$.email_type") AS email_type,
        JSON_VALUE(event_properties, "$.email_provider") AS email_provider,
        JSON_VALUE(event_properties, "$.oauth_client_id") AS oauth_client_id,
        JSON_VALUE(event_properties, "$.connect_device_flow") AS connect_device_flow,
        JSON_VALUE(event_properties, "$.connect_device_os") AS connect_device_os,
        JSON_VALUE(user_properties, "$.sync_device_count") AS sync_device_count,
        JSON_VALUE(user_properties, "$.sync_active_devices_day") AS sync_active_devices_day,
        JSON_VALUE(user_properties, "$.sync_active_devices_week") AS sync_active_devices_week,
        JSON_VALUE(user_properties, "$.sync_active_devices_month") AS sync_active_devices_month,
        JSON_VALUE(event_properties, "$.email_sender") AS email_sender,
        JSON_VALUE(event_properties, "$.email_service") AS email_service,
        JSON_VALUE(event_properties, "$.email_template") AS email_template,
        JSON_VALUE(event_properties, "$.email_version") AS email_version,
      FROM
        unioned
    useLegacySql: false

status:
  errorResult:
    location: query
    message: Field name device_id does not exist in STRUCT<op STRING, os_version STRING,
      event_properties STRING, ...> at [17:24]
    reason: invalidQuery
  errors:
  - location: query
    message: Field name device_id does not exist in STRUCT<op STRING, os_version STRING,
      event_properties STRING, ...> at [17:24]
    reason: invalidQuery
  state: DONE
```

**Comment by kik-kik** [comment]
Hey @whd ,

Apologies, I only saw your comment now. Is this still an issue?

Thanks,
Kris

**Pull request merged**